### PR TITLE
Authority validation strategy

### LINKED ISSUE #226: DiscoveryPolicy: Issuer/authority name validation is too strict
The discovery document I consume exposes an issuer name in the form of `https://domain:443/oauth2/`. 

The validation fails when the string comparison is based on `HttpClient` base URL `https://domain/oauth2`. When comparison occurs, the `HttpClient` base URL seems to be simplified. I didn't verified if it's a behavior of `HttpClient` or `DiscoveryEndpoint` (maybe a combination of both), but I cannot currently force the compared URL to `https://domain:443/oauth2/`) without setting explicitly `DiscoveryPolicy.Authority` . I think it would be inconvenient and fragile because I have to have two configuration values (authority base URL *and* a reference string to compare) instead of one (the authority base URL).

I would like to propose a PR with following changes:

- Add a property `AuthorityNameComparisonStrategy`:`IAuthorityNameComparisonStrategy` to `DiscoveryPolicy`.
- Mark the property `DiscoveryPolicy.AuthorityNameComparison` obsolete.
- Implement `StrictAuthorityNameComparisonStrategy` with the current default (`StringComparison.Ordinal`) and use it as default strategy.
- Implement `UrlAuthorityNameComparisonStrategy` to provide a way to validate equivalence between two URLs (based on `URI` equality method plus trailing slash removal).
- Add method `DiscoveryDocumentResponse.ValidateIssuerName(string issuer, string authority, IAuthorityNameComparisonStrategy strategy)`
- Update `DiscoveryDocumentResponse.Validate(DiscoveryPolicy)` to use the strategy for issuer name validation.

Are you ok with this ?

### PR DESCRIPTION
**Pull request by eric-b**
Fixes #226

### CODE REVIEW

File: src/Client/AuthorityUrlValidationStrategy.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+{
+    /// <summary>
+    /// Implementation of <see cref="IAuthorityValidationStrategy"/> based on <see cref="Uri"/> equality.
+    /// </summary>
+    public sealed class AuthorityUrlValidationStrategy : IAuthorityValidationStrategy
+    {
+        /// <summary>
+        /// Singleton instance.
+        /// </summary>
+        public static AuthorityUrlValidationStrategy Instance = new AuthorityUrlValidationStrategy();
```
Comment (leastprivilege) [inline src/Client/AuthorityUrlValidationStrategy.cs]: Is there a good reason to use the "Instance" pattern here?

File: src/Client/AuthorityUrlValidationStrategy.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+        private AuthorityUrlValidationStrategy()
+        {
+            
+        }
+
+        public AuthorityValidationResult IsIssuerNameValid(string issuerName, string expectedAuthority)
+        {
+            if (!Uri.TryCreate(expectedAuthority.RemoveTrailingSlash(), UriKind.Absolute, out var expectedAuthorityUrl))
```
Comment (leastprivilege) [inline src/Client/AuthorityUrlValidationStrategy.cs:17]: Please use { } here...

File: src/Client/AuthorityUrlValidationStrategy.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+                throw new ArgumentOutOfRangeException("Authority must be a valid URL.", nameof(expectedAuthority));
+
+            if (string.IsNullOrWhiteSpace(issuerName)) return AuthorityValidationResult.CreateError("Issuer name is missing");
+
+            if (!Uri.TryCreate(issuerName.RemoveTrailingSlash(), UriKind.Absolute, out var issuerUrl))
```
Comment (leastprivilege) [inline src/Client/AuthorityUrlValidationStrategy.cs:27]: ..and here

File: src/Client/AuthorityUrlValidationStrategy.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+                return AuthorityValidationResult.CreateError("Issuer name is not a valid URL");
+
+            if (expectedAuthorityUrl.Equals(issuerUrl))
```
Comment (leastprivilege) [inline src/Client/AuthorityUrlValidationStrategy.cs:32]: ..and here

File: src/Client/AuthorityUrlValidationStrategy.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                return AuthorityValidationResult.SuccessResult;
+
+            return AuthorityValidationResult.CreateError("Issuer name does not match authority: " + issuerName);
+        }
+
+        public AuthorityValidationResult IsEndpointValid(string endpoint, IEnumerable<string> allowedAuthorities)
+        {
+            if (string.IsNullOrEmpty(endpoint))
+                return AuthorityValidationResult.CreateError("endpoint is empty");
```
Comment (leastprivilege) [inline src/Client/AuthorityUrlValidationStrategy.cs:44]: IOW - I really don't like omission of curly braces.. I could do it myself - or do you want to change that?

File: src/Client/DiscoveryPolicy.cs [hunk 6]
```diff
@@ -19,7 +22,25 @@ public class DiscoveryPolicy
...
+                _authorityNameComparison = value;
+                AuthorityValidationStrategy = new StringComparisonAuthorityValidationStrategy(value);
+            }
+        }
+
+        /// <summary>
+        /// Strategy used to validate issuer name and endpoints based on expected authority.
+        /// Defaults to <see cref="StringComparisonAuthorityValidationStrategy.Default"/>.
+        /// </summary>
+        public IAuthorityValidationStrategy AuthorityValidationStrategy { get; set; } = new StringComparisonAuthorityValidationStrategy(defaultAuthorityNameComparison);
```
Comment (leastprivilege) [inline src/Client/DiscoveryPolicy.cs]: Here you are newing up the strategy - instead of using Instance. I'd prefer consistency.

**Comment by leastprivilege** [comment]
I left a couple of comments.

btw - 4.x is a breaking release. IOW - no need to obsolete or keep old stuff.

File: src/Client/DiscoveryPolicy.cs (hunk 6 above)
Comment (eric-b) [reply src/Client/DiscoveryPolicy.cs]: I used "Instance" with the non parameterizable class AuthorityUrlValidationStrategy.
The class StringComparisonAuthorityValidationStrategy is parameterizable with StringComparison. IMHO it's consistent, but it maybe is a matter of taste.

File: src/Client/AuthorityUrlValidationStrategy.cs (hunk 1 above)
Comment (eric-b) [reply src/Client/AuthorityUrlValidationStrategy.cs]: My reasoning about that is that creating different instances brings no benefit over singleton "Instance" pattern (no parameterization and no state). Probably not a "good" reason. I removed that.

**Comment by eric-b** [comment]
I think I addressed all your comments except on obsolete stuff.
Policy about obsolete is different on other projects: Would not it be better to keep obsolete on a major version immediately superior to accompany the adaptation from previous major version?
I can remove that if you will.

**Comment by leastprivilege** [comment]
4.0 is a breaking change already from 3.x - so please design it in a way that you don't need to keep backwards compat.

**Comment by eric-b** [comment]
Ok I will make the changes before the end of the week.

**Comment by leastprivilege** [comment]
thanks!

**Comment by eric-b** [comment]
I made the requested changes.
There was a bug introduced in the new URL strategy. I updated tests in consequence.
I made the URL comparison the default. Feel free to change it to the historic string comparison strategy (there is an internal static readonly property `DiscoveryPolicy.DefaultAuthorityValidationStrategy`).

**Comment by leastprivilege** [comment]
looks good. thanks!

**Pull request merged**